bitcoin/src/bip32.rs

```diff
@@ -373,103 +620,209 @@ impl<'a> Iterator for DerivationPathIterator<'a> {
-    /// Create a new [DerivationPath] that is a child of this one.
-    pub fn child(&self, cn: ChildNumber) -> DerivationPath {
-        let mut path = self.0.clone();
-        path.push(cn);
-        DerivationPath(path)
+        /// Get an [Iterator] over the unhardened children of this [DerivationPath].
+        pub fn normal_children(&self) -> DerivationPathIterator {
+            DerivationPathIterator::start_from(self, ChildNumber::ZERO_NORMAL)
+        }
+
+        /// Get an [Iterator] over the hardened children of this [DerivationPath].
+        pub fn hardened_children(&self) -> DerivationPathIterator {
+            DerivationPathIterator::start_from(self, ChildNumber::ZERO_HARDENED)
+        }
+
+        /// Concatenate `self` with `path` and return the resulting new path.
+        ///
+        /// ```
+        /// use bitcoin::bip32::{DerivationPath, ChildNumber};
+        /// use std::str::FromStr;
+        ///
+        /// let base = DerivationPath::from_str("m/42").unwrap();
+        ///
+        /// let deriv_1 = base.extend(DerivationPath::from_str("0/1").unwrap());
+        /// let deriv_2 = base.extend(&[
+        ///     ChildNumber::ZERO_NORMAL,
+        ///     ChildNumber::ONE_NORMAL
+        /// ]);
+        ///
+        /// assert_eq!(deriv_1, deriv_2);
+        /// ```
+        pub fn extend<T: AsRef<[ChildNumber]>>(&self, path: T) -> Self {
+            self.into_iter().chain(path.as_ref()).collect()
+        }
+
+        /// Returns the derivation path as a vector of u32 integers.
+        /// Unhardened elements are copied as is.
+        /// 0x80000000 is added to the hardened elements.
+        ///
+        /// ```
+        /// use bitcoin::bip32::DerivationPath;
+        /// use std::str::FromStr;
+        ///
+        /// let path = DerivationPath::from_str("m/84'/0'/0'/0/1").unwrap();
+        /// const HARDENED: u32 = 0x80000000;
+        /// assert_eq!(path.into_u32_vec(), vec![84 + HARDENED, HARDENED, HARDENED, 0, 1]);
+        /// ```
+        pub fn into_u32_vec(&self) -> Vec<u32> { self.into_iter().map(|&el| el.into()).collect() }
+
+        /// Creates a derivation path from a slice of u32s.
+        /// ```
+        /// use bitcoin::bip32::DerivationPath;
+        ///
+        /// const HARDENED: u32 = 0x80000000;
+        /// let expected = vec![84 + HARDENED, HARDENED, HARDENED, 0, 1];
+        /// let path = DerivationPath::from_u32_slice(expected.as_slice());
+        /// assert_eq!(path.into_u32_vec(), expected);
+        /// ```
+        pub fn from_u32_slice(numbers: &[u32]) -> Self {
+            numbers.iter().map(|&n| ChildNumber::from(n)).collect()
+        }
     }
 
-    /// Convert into a [DerivationPath] that is a child of this one.
-    pub fn into_child(self, cn: ChildNumber) -> DerivationPath {
-        let mut path = self.0;
-        path.push(cn);
-        DerivationPath(path)
+    impl AsRef<[ChildNumber]> for DerivationPath {
+        fn as_ref(&self) -> &[ChildNumber] { &self.0 }
```

Comment:
This leaks implementation detail, so we need to consider removing it.

Reply:
Doing nothing because of #3107 discussion